### PR TITLE
New Entry : Regular Expressions

### PR DESCRIPTION
**Pull request by SSwiniarski**
### Description

Added a general Regular Expressions Entry, and a Java Regular expressions entry. From Target Keyword List.

### Type of Change

- Adding a new entry
- Editing an existing entry (fixing a typo, bug, issues, etc)

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

File: content/python/concepts/regex/regex.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ---
 Title: 'Regular Expressions'
-Description: 'Regex, short for Regular Expressions, is a powerful system for searching text. Regular expressions are implemented in a variety of languages including java and python. In Python, regular expressions are available through the re module: python import re  The following are some of the regular expression operations:'
+Description: 'Regex, short for Regular Expressions, is a powerful system for searching text.'
```
Comment (KTom101) [inline content/python/concepts/regex/regex.md:3]: We should lowercase "regular expressions" to maintain the consistency we have.

```suggestion
Description: 'Regex, short for regular expressions, is a powerful system for searching text.'
```

File: content/python/concepts/regex/regex.md [hunk 2]
```diff
@@ -15,7 +15,7 @@ CatalogContent:
   - 'paths/front-end-engineer-career-path'
 ---
 
-Regex, short for Regular Expressions, is a powerful system for searching text. Regular expressions are implemented in a variety of languages including java and python.
+Regex, short for [Regular Expressions](https://www.codecademy.com/resources/docs/regular-expressions), is a powerful system for searching text. Regular expressions are implemented in a variety of languages including java and python.
```
Comment (KTom101) [inline content/python/concepts/regex/regex.md:18]:
```suggestion
Regex, short for [regular expressions](https://www.codecademy.com/resources/docs/regular-expressions), is a powerful system for searching text. Regular expressions are implemented in a variety of languages including java and python.
```

File: content/python/concepts/regex/terms/findall/findall.md [hunk 3]
```diff
@@ -22,7 +22,7 @@ Iterates over a string to find a subset of characters that match a specified pat
 re.findall(<pattern>, string)
 ```
 
-where `<pattern>` can include any of the following:
+where `<pattern>` is a [regular expression](https://www.codecademy.com/resources/docs/regular-expressions) that can include any of the following:
```
Comment (KTom101) [inline content/python/concepts/regex/terms/findall/findall.md:25]:
```suggestion
Where `<pattern>` is a [regular expression](https://www.codecademy.com/resources/docs/regular-expressions) that can include any of the following:
```

File: content/general/concepts/regular-expressions/regular-expressions.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+|      `^`      | Matches the start of text, or the start of a line, depending on the settings of the regex engine.   | `^a` only matches the first "a" in "a match is made"  |
+|      `$`      | Matches the end of text, or the end of a line, depending on the settings of the regex engine.       | `!$` only matches the second "!" in "Hello! World!"   |
+|      `.`      | Matches any one character.                                                                          | `.` matches "c","a" and "t" in "cat"                  |
+|      `*`      | Matches the preceding element zero or more times.                                                   | `co*t` matches "ct", "cot" and "coot"                 |
+|      `+`      | Matches the preceding element one or more times.                                                    | `co+t` matches "cot","coot" but not "ct"              |
+|      `?`      | Matches the preceding element zero or one time.                                                     | `co?t` matches "ct","cot" but not "coot"              |
```
Comment (KTom101) [inline content/general/concepts/regular-expressions/regular-expressions.md:33]:
```suggestion
|      `?`      | Matches the preceding element zero or one time.                                                     | `co?t` matches "ct", "cot" but not "coot"              |
```

File: content/general/concepts/regular-expressions/regular-expressions.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+|     `{n}`     | Matches the preceding element exactly `n` times.                                                    | `co{2}t` matches "coot" but not "cot" or "cooot"      |
+|   `{,max}`    | Matches the preceding element zero to `max` times.                                                  | `co{,2]t` matches "ct","cot","coot" but not "cooot"   |
```
Comment (KTom101) [inline content/general/concepts/regular-expressions/regular-expressions.md:35]:
```suggestion
|   `{,max}`    | Matches the preceding element zero to `max` times.                                                  | `co{,2]t` matches "ct", "cot", "coot" but not "cooot"   |
```

File: content/general/concepts/regular-expressions/regular-expressions.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+Regular expressions use the following metacharacters and sequences for pattern matching:
+
+| Metacharacter | Description                                                                                         | Example                                               |
+| :-----------: | --------------------------------------------------------------------------------------------------- | ----------------------------------------------------- |
+|      `^`      | Matches the start of text, or the start of a line, depending on the settings of the regex engine.   | `^a` only matches the first "a" in "a match is made"  |
+|      `$`      | Matches the end of text, or the end of a line, depending on the settings of the regex engine.       | `!$` only matches the second "!" in "Hello! World!"   |
+|      `.`      | Matches any one character.                                                                          | `.` matches "c","a" and "t" in "cat"                  |
```
Comment (KTom101) [inline content/general/concepts/regular-expressions/regular-expressions.md:30]: Add spacing after commas; I find it easier to read.
```suggestion
|      `.`      | Matches any one character.                                                                          | `.` matches "c", "a" and "t" in "cat"                  |
```

File: content/general/concepts/regular-expressions/regular-expressions.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+|      `*`      | Matches the preceding element zero or more times.                                                   | `co*t` matches "ct", "cot" and "coot"                 |
+|      `+`      | Matches the preceding element one or more times.                                                    | `co+t` matches "cot","coot" but not "ct"              |
```
Comment (KTom101) [inline content/general/concepts/regular-expressions/regular-expressions.md:32]:
```suggestion
|      `+`      | Matches the preceding element one or more times.                                                    | `co+t` matches "cot", "coot" but not "ct"              |
```

File: content/general/concepts/regular-expressions/regular-expressions.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+Implementation in each programming language is generally through functions and methods that accept some text and a regular expression pattern, and return a result based on what the regular expression pattern matched in the text.
+
+## Regular Expression Patterns
+
+A regular expression pattern is a string made up of normal characters mixed with metacharacters that have special meanings defining how characters should be matched. For instance, the pattern "h" by itself would match the single letter "h", where if we add the metacharacter ".", which matches any one character, then "h." would match sequences such as "ha","ho","h5" and "h!".
```
Comment (KTom101) [inline content/general/concepts/regular-expressions/regular-expressions.md:22]:
```suggestion
A regular expression pattern is a string made up of normal characters mixed with metacharacters that have special meanings defining how characters should be matched. For instance, the pattern "h" by itself would match the single letter "h", whereas if we add the metacharacter ".", which matches any one character, then "h." would match sequences such as "ha", "ho", "h5" and "h!".
```

File: content/java/concepts/regular-expressions/regular-expressions.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
+---
+Title: 'Regular Expressions'
+Description: 'Regular expressions are a language used for pattern-matching text content, and they are implemented in Java through the Pattern and Matcher classes.'
+Subjects:
+  - 'Code Foundations'
+  - 'Computer Science'
+Tags:
+  - 'Strings'
```
Comment (KTom101) [inline content/java/concepts/regular-expressions/regular-expressions.md:8]:
```suggestion
  - 'Regular Expressions'
  - 'Strings'
```

**Review by KTom101 — CHANGES_REQUESTED** [review]
@SSwiniarski I've added the first review for this.

**Comment by SSwiniarski** [comment]
@KTom101, I implemented your suggestions.

**Review by KTom101 — APPROVED** [review]
@SSwiniarski Looks good! It's ready for the next review.

**Comment by KTom101** [comment]
@yangc95, @Dusch4593 This is ready for a second review when you're ready.

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@SSwiniarski While reviewing, I noticed that you used the following link in the most of the files when referring to "regular expressions":

`https://www.codecademy.com/resources/docs/regular-expressions`

Was there supposed to be a topic entry about Regex?

**Comment by SSwiniarski** [comment]
@Dusch4593, I completed your suggestions. Thanks for catching that link error.

**Review by Dusch4593 — APPROVED** [review]
@SSwiniarski LGTM! 👍🏻

**Comment by Dusch4593** [comment]
@SSwiniarski About to merge but I made a change in 1a3918451045e46fef6a2bb79e3372c2c6483512 where I fixed the broken link to the General regex entry and removed the duplicate link.

**Pull request merged**